gordo_components/builder/azure_utils.py

```diff
@@ -0,0 +1,470 @@
+
+
+def _datetime_to_ms_since_epoch(dt) -> int:
+    """
+    Convert datetime to milliseconds since Unix epoch (UTC)
+
+    Parameters
+    ----------
+    dt: datetime.datetime
+        Timestamp to convert (can be timezone aware or naive).
+
+    Returns
+    -------
+    dt: int
+        Timestamp as milliseconds since Unix epoch
+
+    Example
+    -------
+    >>> dt = datetime(1970, 1, 1, 0, 0)
+    >>> _datetime_to_ms_since_epoch(dt)
+    0
+    """
+    epoch = datetime.utcfromtimestamp(0).replace(tzinfo=dt.tzinfo)
+    dt = dt.astimezone(UTC) if dt.tzinfo else dt.replace(tzinfo=dt.tzinfo)
+    return round((dt - epoch).total_seconds() * 1000.0)
+
+
+def get_batch_kwargs(
+    metadata: dict, timestamp: datetime = datetime.now(tz=UTC), step: int = 0
+) -> dict:
+    """
+    Create flat lists of MLflow logging entities from multilevel dictionary
+
+    For more information, see the mlflow docs:
+    https://www.mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.log_batch
+
+    Parameters
+    ----------
+    metadata: dict
+        Multi-level dictionary with (key, value) loggables to parse.
+    timestamp: datetime
+        Timezone aware timestamp of time to associate with all metrics logged.
+    step: int
+        Training step to associate with all metrics.
+
+    Returns
+    -------
+    batch_kwargs: dict
+        Dict with `metrics` and `param_list` lists for passing to
+        `mlflow.tracking.MlflowClient.log_batch`. The metrics key containes a list of
+        Mlflow `Metric` instances, and the param_list key contains a list of `Param`
+        instances.
+    """
+
+    # Set timestamp if not passed
+    if not timestamp or timestamp is None:
+        timestamp = datetime.now(UTC)
+
+    # Verify timestamp is timezone aware
+    if not timestamp.tzinfo:
+        raise ValueError(
+            f"Timestamp ({timestamp} needs to include timezone information"
+        )
+
+    # Validate step is passed as an int
+    if type(step) is not int:
+        raise ValueError(
+            f"`step` argument must be of type '{str(int)}' not '{type(step)}'"
+        )
+
+    ts_int = _datetime_to_ms_since_epoch(timestamp)
+
+    def to_str(x: Union[datetime, Any]) -> str:
+        """Format datetimes as strings else strict cast as string"""
+        if type(x) is datetime:
+            return x.strftime("%Y-%m-%d %H:%M:%S.%f+%z")
+        else:
+            return str(x)
+
+    def get_params(d: dict, keys: List[str], prefix: str = "") -> List[Param]:
+        """Get list of single summary metrics in dictionary defined by keys"""
+        return [Param(k, to_str(d[k])) for k in keys if k in d]
+
+    def get_metrics(d: dict, keys: List[str], prefix: str = "") -> List[Metric]:
+        """Get list of single summary metrics in dictionary defined by keys"""
+        return [Metric(k, float(d[k]), ts_int, 0) for k in keys if k in d]
+
+    metric_list: List[Metric] = list()
```

Comment:
Could you just call the get_metrics to initialize? Seems a bit unneccessarry to first have an empty list, when you have a method that is called in the next step?